specifyweb/specify/tree_views.py

```diff
@@ -342,4 +343,361 @@ def perm_target(tree):
+        # - the default rank is greater than the target rank
+        # - the default rank is less than the current next rank from the target rank
+        if can_use_default_rank_id:
+            default_rank_id = default_tree_ranks[new_rank_name.upper()]
+
+            # Check if the default rank ID is not already used
+            is_default_rank_id_unused = default_rank_id not in rank_ids
+
+            # Check if the default rank ID can be logically placed in the hierarchy
+            is_placement_valid = (
+                is_tree_def_items_empty 
+                or (is_new_rank_first and default_rank_id < next_rank_id) 
+                or (is_new_rank_last and default_rank_id > target_rank_id)
+                or (default_rank_id > target_rank_id and default_rank_id < next_rank_id)
+            )
+
+            if is_default_rank_id_unused and is_placement_valid:
+                new_rank_id = default_rank_id
+
+        # Set the new rank id if a default rank id is not available
+        if new_rank_id is None:
+            # If this is the first rank, create ...
+            if is_tree_def_items_empty:
+                new_fields_dict['rankid'] = rank_increment
+
+            # If there are no ranks higher than the target rank, then add the new rank to the end of the hierarchy
+            elif is_new_rank_first:
+                new_rank_id = int((next_rank_id - target_rank_id) / 2) + target_rank_idx
+                max_rank_id = rank_ids[-1]
+                new_fields_dict['rankid'] = max_rank_id + rank_increment
+
+            # If there are no ranks lower than the target rank, then add the new rank to the top of the hierarchy
+            elif is_new_rank_last:
+                min_rank_id = rank_ids[0]
+                new_fields_dict['rankid'] = min_rank_id + rank_increment # TODO: checkout
+
+            # If the new rank is being placed somewhere in the middle of the heirarchy
+            else:
+                new_rank_id = int((next_rank_id - target_rank_id) / 2) + target_rank_id
+                if next_rank_id - target_rank_id < 1:
+                    raise Exception(f"Can't add rank id between {new_rank_id} and {target_rank_id}")
+                
+        # Create and save the new TreeDefItem record
+        new_fields_dict['rankid'] = new_rank_id
+        tree_def_item = tree_def_item_model.objects.create(**new_fields_dict)
+        tree_def_item.save()
+
+        # Create a new tree def item
+        tree_def_item_model.objects.create(id=new_rank_id, name=new_rank_name, title=new_rank_title)
+
+        # Regenerate full names
+        tree_extras.set_fullnames(tree_def_model, null_only=False, node_number_range=None)
+
+        logger.info(f"Added new tree rank: {new_rank_name} with ID: {new_rank_id}")
+        return HttpResponse("Success")
+    
+    except Exception as e:
+        logger.error(f"Error in adding tree rank: {str(e)}")
+        return HttpResponseServerError(f"Failed: {str(e)}")
```

Comment:
do we normally wrap all api endpoints in try/catch?
doesn't HttpResponseServerError (or something like that) already get returned to the front-end if there is an uncaught error?
thus, what's the benefit of explicitly catching and re-throwing the error?